ci/upload_audit_events_s3.py

```diff
@@ -0,0 +1,179 @@
+            return
+
+        with requests.Session() as s:
+            s.headers["Authorization"] = f"Bearer {self.token}"
+            url = urljoin(self.CF_API_URL, f"v3/{entity_path}/{entity_guid}")
+            response = s.get(url)
+
+            if response.status_code == 404:
+                return
+
+            data = response.json()
+            return data["name"]
+
+    def get_audit_logs(self, start, end):
+        audit_logs = []
+
+        with requests.Session() as s:
+            s.headers["Authorization"] = f"Bearer {self.token}"
+            params = {
+                "created_ats[gt]": str(start),
+                "created_ats[lt]": str(end),
+                "order_by": "created_at",
+            }
+            url = urljoin(self.CF_API_URL, "/v3/audit_events")
+
+            first_response = s.get(url, params=params)
+            data = first_response.json()
+            audit_logs.extend(data["resources"])
+
+            while data["pagination"]["next"] is not None:
+                data = s.get(data["pagination"]["next"]["href"]).json()
+                audit_logs.extend(data["resources"])
+
+        return audit_logs
+
+    def transform_audit_event(self, audit_event):
+        # remove "links" property from event
+        transformed_event = {k: v for k, v in audit_event.items() if k not in ["links"]}
+
+        organization = audit_event.get("organization", {})
+        if organization:
+            organization_name = self.get_cf_entity_name(
+                "organizations",
+                organization.get("guid", None),
+            )
+            if organization_name:
+                transformed_event["organization_name"] = organization_name
+
+        space = audit_event.get("space", {})
+        if space:
+            space_name = self.get_cf_entity_name(
+                "spaces",
+                space.get("guid", None),
+            )
+            if space_name:
+                transformed_event["space_name"] = space_name
+
+        return transformed_event
+
+    # Upload a batch of audit events to S3 as a single object
+    def put_audit_events_to_s3(self, object_name, audit_events):
+        body = "\n".join(
+            [
+                json.dumps(self.transform_audit_event(audit_event))
+                for audit_event in audit_events
+            ]
+        )
+        s3_client.put_object(
+            Bucket=self.bucket_name,
+            Key=object_name,
+            Body=body,
+            ContentType="application/json",
+        )
+
+    def update_latest_stamp_in_s3(self, latest_timestamp):
+        data = latest_timestamp
+        s3_client.put_object(
+            Bucket=self.bucket_name,
+            Key=timestamp_key,
+            Body=data,
+        )
+
+    def get_start_end_time(self, now):
+        fifteen_minutes_ago = now - timedelta(minutes=15)
+        start_time = fifteen_minutes_ago.strftime("%Y-%m-%dT%H:%M:%SZ")
+        end_time = now.strftime("%Y-%m-%dT%H:%M:%SZ")
```

Comment:
I don't know if this will handle timestamps correctly, or just pretend that the Pacific Time stamps are actually UTC, which will cause problems with DST and if we ever move this or run it from more than one timezone

Reply:
Since `now` is initialized in UTC time, won't it handled things correctly?

Reply:
oh, yeah. I didn't catch that because `now` was defined elsewhere. That fixes that.